File: src/data/repositories/SupabaseLearnerRepository.ts
import { ILearnerRepository, AnswerAttemptRecord, PlacementAttemptData } from '../../domain/repositories/ILearnerRepository';
import { LearnerProfile } from '../../domain/entities/LearnerProfile';
import { SkillScore } from '../../domain/entities/SkillScore';
import { InMemoryLearnerRepository } from './InMemoryLearnerRepository';
import { supabase, isSupabaseConfigured } from '../../core/config/supabase';

export class SupabaseLearnerRepository implements ILearnerRepository {
  private fallbackRepo = new InMemoryLearnerRepository();

  async getCurrentProfile(): Promise<LearnerProfile> {
    return this.getProfile();
  }

  async getProfile(): Promise<LearnerProfile> {
    if (!isSupabaseConfigured) {
      return this.fallbackRepo.getProfile();
    }

    try {
      const { data: { user } } = await supabase.auth.getUser();
      if (!user) {
        return this.fallbackRepo.getProfile();
      }

      // Fetch user profile from public.profiles table
      let { data: userProfileData } = await supabase
        .from('profiles')
        .select('id, first_name, last_name, display_name, email, avatar_url, created_at, updated_at')
        .eq('id', user.id)
        .maybeSingle();

      // If public.profiles row is missing or empty, provision it
      if (!userProfileData || !userProfileData.first_name) {
        const metaFirst = (user.user_metadata?.first_name || '').trim();
        const metaLast = (user.user_metadata?.last_name || '').trim();
        const resolvedFirst = metaFirst || (user.email ? user.email.split('@')[0] : 'Foydalanuvchi');
        const resolvedDisplay = metaLast ? `${resolvedFirst} ${metaLast}` : resolvedFirst;

        const { data: provisioned } = await supabase
          .from('profiles')
          .upsert({
            id: user.id,
            first_name: resolvedFirst,
            last_name: metaLast,
            display_name: resolvedDisplay,
            email: user.email || '',
            avatar_url: user.user_metadata?.avatar_url || null,
            updated_at: new Date().toISOString(),
          })
          .select()
          .maybeSingle();

        if (provisioned) {
          userProfileData = provisioned;
        }
      }

      // Fetch learner profile
      const { data: learnerData } = await supabase
        .from('learner_profiles')
        .select('*')
        .eq('user_id', user.id)
        .maybeSingle();

      // Fetch gamification profile
      const { data: gamifyData } = await supabase
        .from('gamification_profiles')
        .select('*')
        .eq('user_id', user.id)
        .maybeSingle();

      // Fetch skill scores
      const { data: scoreRows } = await supabase
        .from('learner_skill_scores')
        .select('*')
        .eq('user_id', user.id);

      // Fetch completed lessons
      const { data: progressRows } = await supabase
        .from('lesson_progress')
        .select('lesson_id, status')
        .eq('user_id', user.id)
        .eq('status', 'completed');

      // Fetch completed reinforcements
      const { data: reinfRows } = await supabase
        .from('reinforcement_attempts')
        .select('reinforcement_node_id, is_correct')
        .eq('user_id', user.id)
        .eq('is_correct', true);

      const scoresByCourse: Record<string, Record<string, SkillScore>> = {};
      if (scoreRows) {
        for (const row of scoreRows) {
          scoresByCourse[row.course_id] = scoresByCourse[row.course_id] || {};
          scoresByCourse[row.course_id][row.skill_id] = {
            skillId: row.skill_id,
            courseId: row.course_id,
            score: row.score,
            lastUpdated: new Date(row.last_updated).getTime(),
            masteryLevel: row.mastery_level,
            isWeakestFocus: row.is_weakest_focus,
          };
        }
      }

      const firstName = userProfileData?.first_name ||
        user.user_metadata?.first_name ||
        (user.email ? user.email.split('@')[0] : 'Foydalanuvchi');
      const lastName = userProfileData?.last_name || user.user_metadata?.last_name || '';

      return {
        id: user.id,
        name: firstName,
        firstName,
        lastName,
        email: user.email || userProfileData?.email || '',
        avatarUrl: userProfileData?.avatar_url || '',
        selectedCourseId: learnerData?.selected_course_id || 'course_math_01',
        goal: learnerData?.goal || 'mastery',
        dailyMinutes: learnerData?.daily_minutes || 15,
        initialLevel: learnerData?.initial_level || 'intermediate',
        xp: gamifyData?.xp || 0,
        streakDays: gamifyData?.streak_days || 1,
        lastActiveDate: gamifyData?.last_activity_date || new Date().toISOString().split('T')[0],
        scoresByCourse,
        completedLessonIds: (progressRows || []).map((r) => r.lesson_id),
        completedNodeIds: [],
        completedReinforcementIds: (reinfRows || []).map((r) => r.reinforcement_node_id),
        createdAt: new Date(learnerData?.created_at || Date.now()).getTime(),
      };
    } catch (err) {
      console.warn('Error fetching Supabase learner profile, using local state:', err);
      return this.fallbackRepo.getProfile();
    }
  }

  async updateProfile(updates: Partial<LearnerProfile>): Promise<LearnerProfile> {
    if (!isSupabaseConfigured) {
      return this.fallbackRepo.updateProfile(updates);
    }

    try {
      const { data: { user } } = await supabase.auth.getUser();
      if (!user) {
        return this.fallbackRepo.updateProfile(updates);
      }

      const currentProfile = await this.getProfile();
      const merged: LearnerProfile = {
        ...currentProfile,
        ...updates,
        scoresByCourse: updates.scoresByCourse || currentProfile.scoresByCourse,
      };

      // Sync public.profiles if name/names/avatar are updated
      if (updates.name !== undefined || updates.firstName !== undefined || updates.lastName !== undefined || updates.avatarUrl !== undefined) {
        const newFirst = updates.firstName || updates.name || currentProfile.firstName || 'Foydalanuvchi';
        const newLast = updates.lastName ?? currentProfile.lastName ?? '';
        const newDisplay = newLast ? `${newFirst} ${newLast}` : newFirst;

        await supabase.from('profiles').upsert({
          id: user.id,
          first_name: newFirst,
          last_name: newLast,
          display_name: newDisplay,
          avatar_url: updates.avatarUrl ?? currentProfile.avatarUrl ?? null,
          updated_at: new Date().toISOString(),
        });
      }

      await supabase.from('learner_profiles').upsert({
        user_id: user.id,
        selected_course_id: merged.selectedCourseId,
        goal: merged.goal,
        daily_minutes: merged.dailyMinutes,
        initial_level: merged.initialLevel,
        updated_at: new Date().toISOString(),
      });

      if (updates.xp !== undefined || updates.streakDays !== undefined || updates.lastActiveDate !== undefined) {
        await supabase.from('gamification_profiles').upsert({
          user_id: user.id,
          xp: merged.xp,
          streak_days: merged.streakDays,
          last_activity_date: merged.lastActiveDate,
          updated_at: new Date().toISOString(),
        });
      }

      return merged;
    } catch (err) {
      console.warn('Error updating Supabase learner profile, using local state:', err);
      return this.fallbackRepo.updateProfile(updates);
    }
  }

  async saveSkillScores(courseId: string, scores: Record<string, SkillScore>): Promise<void> {
    if (!isSupabaseConfigured) {
      return this.fallbackRepo.saveSkillScores(courseId, scores);
    }

    try {
      const { data: { user } } = await supabase.auth.getUser();
      if (!user) {
        return this.fallbackRepo.saveSkillScores(courseId, scores);
      }

      const rows = Object.values(scores).map((s) => ({
        user_id: user.id,
        course_id: courseId,
        skill_id: s.skillId,
        score: s.score,
        mastery_level: s.masteryLevel,
        is_weakest_focus: Boolean(s.isWeakestFocus),
        last_updated: new Date().toISOString(),
      }));

      await supabase.from('learner_skill_scores').upsert(rows, {
        onConflict: 'user_id,course_id,skill_id',
      });
    } catch (err) {
      console.warn('Error saving skill scores to Supabase, saving locally:', err);
      return this.fallbackRepo.saveSkillScores(courseId, scores);
    }
  }

  async savePlacementAttempt(data: PlacementAttemptData): Promise<string> {
    if (!isSupabaseConfigured) {
      return this.fallbackRepo.savePlacementAttempt(data);
    }

    try {
      const { data: { user } } = await supabase.auth.getUser();
      if (!user) {
        return this.fallbackRepo.savePlacementAttempt(data);
      }

      const { data: attemptRow, error: attemptErr } = await supabase
        .from('placement_attempts')
        .insert({
          user_id: user.id,
          course_id: data.courseId,
          score: data.score,
          weakest_skill_id: data.weakestSkillId || null,
          started_at: new Date().toISOString(),
          completed_at: new Date().toISOString(),
        })
        .select('id')
        .single();

      if (attemptErr || !attemptRow) {
        console.warn('Error inserting placement attempt to Supabase:', attemptErr);
        return this.fallbackRepo.savePlacementAttempt(data);
      }

      const attemptId = attemptRow.id;

      if (data.submissions && data.submissions.length > 0) {
        const answerRows = data.submissions.map((sub) => ({
          attempt_id: attemptId,
          user_id: user.id,
          question_id: sub.questionId,
          selected_index: sub.selectedIndex,
          is_correct: sub.isCorrect,
          created_at: new Date().toISOString(),
        }));

        const { error: answersErr } = await supabase
          .from('placement_answers')
          .insert(answerRows);

        if (answersErr) {
          console.warn('Error inserting placement answers to Supabase:', answersErr);
        }
      }

      return attemptId;
    } catch (err) {
      console.warn('Exception saving placement attempt to Supabase:', err);
      return this.fallbackRepo.savePlacementAttempt(data);
    }
  }

  async markLessonCompleted(lessonId: string): Promise<void> {
    await this.fallbackRepo.markLessonCompleted(lessonId);
    if (!isSupabaseConfigured) return;

    try {
      const { data: { user } } = await supabase.auth.getUser();
      if (!user) return;

      await supabase.from('lesson_progress').upsert({
        user_id: user.id,
        lesson_id: lessonId,
        course_id: 'course_math_01',
        status: 'completed',
        progress_percent: 100,
        completed_at: new Date().toISOString(),
        updated_at: new Date().toISOString(),
      }, {
        onConflict: 'user_id,lesson_id',
      });
    } catch (err) {
      console.warn('Error marking lesson completed in Supabase:', err);
    }
  }

  async markNodeCompleted(nodeId: string): Promise<void> {
    return this.fallbackRepo.markNodeCompleted(nodeId);
  }

  async markReinforcementCompleted(reinforcementId: string): Promise<void> {
    await this.fallbackRepo.markReinforcementCompleted(reinforcementId);
    if (!isSupabaseConfigured) return;

    try {
      const { data: { user } } = await supabase.auth.getUser();
      if (!user) return;

      await supabase.from('reinforcement_attempts').upsert({
        user_id: user.id,
        course_id: 'course_math_01',
        skill_id: 'skill_math_functions',
        reinforcement_node_id: reinforcementId,
        before_score: 40,
        after_score: 60,
        is_correct: true,
        xp_awarded: 15,
        created_at: new Date().toISOString(),
      }, {
        onConflict: 'user_id,reinforcement_node_id',
      });
    } catch (err) {
      console.warn('Error marking reinforcement completed in Supabase:', err);
    }
  }

  async recordAnswerAttempt(attempt: Omit<AnswerAttemptRecord, 'id' | 'timestamp'>): Promise<AnswerAttemptRecord> {
    const record = await this.fallbackRepo.recordAnswerAttempt(attempt);
    if (!isSupabaseConfigured) return record;

    try {
      const { data: { user } } = await supabase.auth.getUser();
      if (!user) return record;

      // Invoke server-side RPC if available
      await supabase.rpc('submit_answer_attempt', {
        p_user_id: user.id,
        p_course_id: attempt.courseId,
        p_skill_id: attempt.skillId,
        p_lesson_id: attempt.lessonId || 'general',
        p_question_id: attempt.questionId,
        p_selected_index: attempt.selectedIndex,
        p_selected_text: attempt.selectedAnswer,
      });
    } catch (err) {
      console.warn('Error recording answer attempt to Supabase, logged locally:', err);
    }

    return record;
  }

  async getAnswerAttempts(limit: number = 10): Promise<AnswerAttemptRecord[]> {
    return this.fallbackRepo.getAnswerAttempts(limit);
  }

  async addXp(amount: number, actionIdempotencyKey?: string): Promise<number> {
    return this.fallbackRepo.addXp(amount, actionIdempotencyKey);
  }

  async recordDailyActivity(dateStr?: string): Promise<number> {
    return this.fallbackRepo.recordDailyActivity(dateStr);
  }

  async resetAll(): Promise<LearnerProfile> {
    return this.fallbackRepo.resetAll();
  }
}

export const supabaseLearnerRepository = new SupabaseLearnerRepository();

File: mobile/lib/data/repositories/supabase_learner_repository.dart
import 'package:supabase_flutter/supabase_flutter.dart';
import '../../domain/repositories/i_learner_repository.dart';
import '../../domain/entities/learner_profile.dart';
import '../../domain/entities/skill_score.dart';
import 'in_memory_learner_repository.dart';

class SupabaseLearnerRepository implements ILearnerRepository {
  final InMemoryLearnerRepository _fallbackRepo = InMemoryLearnerRepository();

  @override
  Future<LearnerProfile> getProfile() async {
    try {
      final client = Supabase.instance.client;
      final user = client.auth.currentUser;
      if (user == null) {
        return _fallbackRepo.getProfile();
      }

      // Fetch learner profile
      final learnerData = await client
          .from('learner_profiles')
          .select()
          .eq('user_id', user.id)
          .maybeSingle();

      // Fetch gamification profile
      final gamifyData = await client
          .from('gamification_profiles')
          .select()
          .eq('user_id', user.id)
          .maybeSingle();

      // Fetch skill scores
      final scoreRows = await client
          .from('learner_skill_scores')
          .select()
          .eq('user_id', user.id);

      // Fetch completed lessons
      final progressRows = await client
          .from('lesson_progress')
          .select('lesson_id')
          .eq('user_id', user.id)
          .eq('status', 'completed');

      // Fetch completed reinforcements
      final reinfRows = await client
          .from('reinforcement_attempts')
          .select('reinforcement_node_id')
          .eq('user_id', user.id)
          .eq('is_correct', true);

      final Map<String, Map<String, SkillScore>> scoresByCourse = {};
      for (final row in scoreRows) {
        final courseId = row['course_id'] as String;
        final skillId = row['skill_id'] as String;
        scoresByCourse[courseId] ??= {};
        scoresByCourse[courseId]![skillId] = SkillScore(
          skillId: skillId,
          courseId: courseId,
          score: (row['score'] as num).toInt(),
          lastUpdated: DateTime.parse(row['last_updated']).millisecondsSinceEpoch,
          masteryLevel: row['mastery_level'] ?? 'developing',
          isWeakestFocus: row['is_weakest_focus'] == true,
        );
      }

      final completedLessonIds = (progressRows as List?)
              ?.map((e) => e['lesson_id'] as String)
              .toList() ??
          [];

      final completedReinfIds = (reinfRows as List?)
              ?.map((e) => e['reinforcement_node_id'] as String)
              .toList() ??
          [];

      // Fetch user profile from public.profiles table
      final userProfile = await client
          .from('profiles')
          .select('first_name, last_name, email')
          .eq('id', user.id)
          .maybeSingle();

      final resolvedName = userProfile?['first_name'] as String? ??
          user.userMetadata?['first_name'] as String? ??
          (user.email != null ? user.email!.split('@').first : 'O‘quvchi');

      return LearnerProfile(
        id: user.id,
        name: resolvedName,
        selectedCourseId: learnerData?['selected_course_id'] ?? 'course_math_01',
        goal: OnboardingGoal.mastery,
        dailyMinutes: learnerData?['daily_minutes'] ?? 15,
        initialLevel: InitialLevel.intermediate,
        xp: gamifyData?['xp'] ?? 0,
        streakDays: gamifyData?['streak_days'] ?? 1,
        lastActiveDate: gamifyData?['last_activity_date'] ??
            DateTime.now().toIso8601String().split('T').first,
        scoresByCourse: scoresByCourse,
        completedLessonIds: completedLessonIds,
        completedNodeIds: completedLessonIds,
        completedReinforcementIds: completedReinfIds,
        createdAt: DateTime.now().millisecondsSinceEpoch,
      );
    } catch (_) {
      return _fallbackRepo.getProfile();
    }
  }

  @override
  Future<LearnerProfile> updateProfile(LearnerProfile updated) async {
    try {
      final client = Supabase.instance.client;
      final user = client.auth.currentUser;
      if (user == null) {
        return _fallbackRepo.updateProfile(updated);
      }

      await client.from('learner_profiles').upsert({
        'user_id': user.id,
        'selected_course_id': updated.selectedCourseId,
        'daily_minutes': updated.dailyMinutes,
        'goal': updated.goal.name,
        'initial_level': updated.initialLevel.name,
      });

      return updated;
    } catch (_) {
      return _fallbackRepo.updateProfile(updated);
    }
  }

  @override
  Future<void> saveSkillScores(String courseId, Map<String, SkillScore> scores) async {
    try {
      final client = Supabase.instance.client;
      final user = client.auth.currentUser;
      if (user == null) {
        return _fallbackRepo.saveSkillScores(courseId, scores);
      }

      for (final entry in scores.entries) {
        await client.from('learner_skill_scores').upsert({
          'user_id': user.id,
          'course_id': courseId,
          'skill_id': entry.key,
          'score': entry.value.score,
          'mastery_level': entry.value.masteryLevel,
          'is_weakest_focus': entry.value.isWeakestFocus,
          'last_updated': DateTime.now().toIso8601String(),
        });
      }
    } catch (_) {
      await _fallbackRepo.saveSkillScores(courseId, scores);
    }
  }

  @override
  Future<String> savePlacementAttempt(PlacementAttemptData data) async {
    try {
      final client = Supabase.instance.client;
      final user = client.auth.currentUser;
      if (user == null) {
        return _fallbackRepo.savePlacementAttempt(data);
      }

      final attemptRow = await client
          .from('placement_attempts')
          .insert({
            'user_id': user.id,
            'course_id': data.courseId,
            'score': data.score,
            'weakest_skill_id': data.weakestSkillId,
            'started_at': DateTime.now().toIso8601String(),
            'completed_at': DateTime.now().toIso8601String(),
          })
          .select('id')
          .single();

      final attemptId = attemptRow['id'] as String;

      if (data.submissions.isNotEmpty) {
        final answerRows = data.submissions
            .map((sub) => {
                  'attempt_id': attemptId,
                  'user_id': user.id,
                  'question_id': sub.questionId,
                  'selected_index': sub.selectedIndex,
                  'is_correct': sub.isCorrect,
                  'created_at': DateTime.now().toIso8601String(),
                })
            .toList();

        await client.from('placement_answers').insert(answerRows);
      }

      return attemptId;
    } catch (_) {
      return _fallbackRepo.savePlacementAttempt(data);
    }
  }

  @override
  Future<void> markLessonCompleted(String lessonId) async {
    try {
      final client = Supabase.instance.client;
      final user = client.auth.currentUser;
      if (user == null) {
        return _fallbackRepo.markLessonCompleted(lessonId);
      }

      await client.from('lesson_progress').upsert({
        'user_id': user.id,
        'lesson_id': lessonId,
        'status': 'completed',
        'completed_at': DateTime.now().toIso8601String(),
      });
      await addXp(20, 'lesson_completed_$lessonId');
    } catch (_) {
      await _fallbackRepo.markLessonCompleted(lessonId);
    }
  }

  @override
  Future<void> markNodeCompleted(String nodeId) async {
    await markLessonCompleted(nodeId);
  }

  @override
  Future<void> markReinforcementCompleted(String reinforcementId) async {
    try {
      final client = Supabase.instance.client;
      final user = client.auth.currentUser;
      if (user == null) {
        return _fallbackRepo.markReinforcementCompleted(reinforcementId);
      }

      await client.from('reinforcement_attempts').upsert({
        'user_id': user.id,
        'reinforcement_node_id': reinforcementId,
        'is_correct': true,
        'created_at': DateTime.now().toIso8601String(),
      });
      await addXp(15, 'reinforcement_completed_$reinforcementId');
    } catch (_) {
      await _fallbackRepo.markReinforcementCompleted(reinforcementId);
    }
  }

  @override
  Future<AnswerAttemptRecord> recordAnswerAttempt({
    required String courseId,
    required String skillId,
    required String lessonId,
    required String questionId,
    required int selectedIndex,
    required String selectedAnswer,
    required bool isCorrect,
  }) async {
    try {
      final client = Supabase.instance.client;
      final user = client.auth.currentUser;
      if (user != null) {
        await client.rpc('submit_answer_attempt', params: {
          'p_user_id': user.id,
          'p_course_id': courseId,
          'p_skill_id': skillId,
          'p_lesson_id': lessonId,
          'p_question_id': questionId,
          'p_selected_index': selectedIndex,
          'p_selected_text': selectedAnswer,
        });
      }
      return _fallbackRepo.recordAnswerAttempt(
        courseId: courseId,
        skillId: skillId,
        lessonId: lessonId,
        questionId: questionId,
        selectedIndex: selectedIndex,
        selectedAnswer: selectedAnswer,
        isCorrect: isCorrect,
      );
    } catch (_) {
      return _fallbackRepo.recordAnswerAttempt(
        courseId: courseId,
        skillId: skillId,
        lessonId: lessonId,
        questionId: questionId,
        selectedIndex: selectedIndex,
        selectedAnswer: selectedAnswer,
        isCorrect: isCorrect,
      );
    }
  }

  @override
  Future<List<AnswerAttemptRecord>> getAnswerAttempts([int limit = 10]) async {
    return _fallbackRepo.getAnswerAttempts(limit);
  }

  @override
  Future<int> addXp(int amount, [String? actionIdempotencyKey]) async {
    return _fallbackRepo.addXp(amount, actionIdempotencyKey);
  }

  @override
  Future<int> recordDailyActivity([String? dateStr]) async {
    return _fallbackRepo.recordDailyActivity(dateStr);
  }

  @override
  Future<LearnerProfile> resetAll() async {
    return _fallbackRepo.resetAll();
  }
}
